feat: modularize onOtherActorDeath and onOtherActorFleeing

LordMidas opened the pull request:
- Add four new morale check types for fleeing/dying allies/enemies.
- Modularize the onOtherActorDeath and onOtherActorFleeing functions to allow fine control on who the morale checks are triggered for and what the difficulty is.
- Add functionality to allow fleeing actors to trigger morale check for enemies.

For reviewers:
In addition to reviewing my code, please view the vanilla `actor.onOtherActorDeath` and `actor.onOtherActorFleeing` as well as find the places in vanilla where these functions are called and double check that my implementation does not change any vanilla behavior.

Darxo commented on mod_modular_vanilla/hooks/entity/tactical/actor.nut:
```diff
@@ -89,6 +132,150 @@
+		// negative morale check dropped this character to fleeing
+		if (_change < 0 && ret && this.getMoraleState() == ::Const.MoraleState.Fleeing)
+		{
+			// vanilla only calls this for allies, but in MV we call it for enemies too
```
```suggestion
			// vanilla only calls this for allies from the same faction, but in MV we call it for all entities of other factions too
```

Darxo commented on mod_modular_vanilla/hooks/entity/tactical/actor.nut:
```diff
@@ -89,6 +132,150 @@
+
+	// MV: Changed
+	// Part of actor.onOtherActorFleeing modularization
+	// add call to actor.onOtherActorFleeing even for enemies (vanilla has only for same faction allies)
```
```suggestion
	// add call to actor.onOtherActorFleeing for all other factions (vanilla has only for same faction allies)
``````

LordMidas commented on mod_modular_vanilla/hooks/entity/tactical/actor.nut:
```diff
@@ -89,6 +132,150 @@
 
+	// MV: Added
+	// Part of actor.onOtherActorDeath modularization
+	q.MV_isCausingMoraleCheckOnDeath <- function( _killer, _skill )
```
Yeah I don't like it either. I only kept it because of the vanilla `_victim.getXPValue() <= 1` condition which doesn't seem to fit inside the `MV_isCausingMoraleCheckOnDeathFor` function e.g. if someone overwrites that then we lose this condition. Perhaps this can be made into a separate Const boolean instead e.g. ```::Const.Morale.MV_MoraleCheckOnDeathXPValueMin` and then if someone wants to remove this condition he could set that to `99999999` and then we can simply include this in the default function and get rid of this extra function?

LordMidas committed:
feat: add surround morale check to modularized morale check handling

LordMidas committed:
feat: modularize onOtherActorDeath and onOtherActorFleeing

- Add four new morale check types for fleeing/dying allies/enemies.
- Modularize the onOtherActorDeath and onOtherActorFleeing functions to allow fine control on who the morale checks are triggered for and what the difficulty is.
- Add functionality to allow fleeing actors to trigger morale check for enemies.

LordMidas commented on mod_modular_vanilla/hooks/entity/tactical/actor.nut:
```diff
@@ -89,6 +132,150 @@
+	{
+		if (_victim.getFaction() == this.getFaction())
+		{
+			return this.Const.Morale.AllyKilledBaseDifficulty - _victim.getXPValue() * this.Const.Morale.AllyKilledXPMult + this.Math.pow(_victim.getTile().getDistanceTo(this.getTile()), this.Const.Morale.AllyKilledDistancePow);
```
I will clean that up eventually in a larger code cleanup in Modular Vanilla. For now, in my opinion, it is just easier to copy paste vanilla code and adjust - adding comments will just pollute it.

LordMidas commented:
I have rewritten and restructured the approach. Now it is much more streamlined with a couple of centralized functions which handle the morale check validity and its difficulty calculation. Please review again.
